Add metrics-tomcat-jmxproxy and fix deprecated codecoverage

luyseyal opened the pull request:
Add metrics-tomcat-jmxproxy and fix deprecated codecoverage

luyseyal commented:
FYI @majormoses

majormoses commented:
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+         short: '-a USERNAME',
+         long: '--account USERNAME',
+         default: 'admin',
+         required: true
```
Indeed, I had the same thought when I cribbed this code from the other scripts in this repo. I just wanted to be consistent with them. I'm happy to drop the default but think it should be done for all the scripts, not just this one.

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+         description: 'Tomcat User password',
+         short: '-p PASSWORD',
+         long: '--password PASSWORD',
+         default: 'password',
```
No, it's not the default, fortunately! The default is no users enabled so for jmxproxy to work, you have to add a user with the role `manager-jmx` to tomcat-users.xml. Also, [Tomcat has a valve to turn off access](https://tomcat.apache.org/tomcat-8.5-doc/manager-howto.html#Introduction) to things like manager, specifically.

BTW, if you're accustomed to turning off management, you really should check out what's available for monitoring. `jmxproxy` is nice because it is just plaintext.
https://wiki.apache.org/tomcat/FAQ/Monitoring
https://tomcat.apache.org/tomcat-8.5-doc/manager-howto.html#Using_the_JMX_Proxy_Servlet

majormoses replied:
if its not the default, I see no value in specifying this default. let's remove this.

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+         short: '-p PASSWORD',
+         long: '--password PASSWORD',
+         default: 'password',
+         required: true
```
Indeed, I had the same thought when I cribbed this code from the other scripts in this repo. I just wanted to be consistent with them. I'm happy to drop the default but think it should be done for all the scripts, not just this one.

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+         short: '-l VALUE',
+         long: '--attribute VALUE',
+         default: 'HeapMemoryUsage',
+         required: true
```
Indeed, I had the same thought when I cribbed this code from the other scripts in this repo. I just wanted to be consistent with them. I'm happy to drop the default but think it should be done for all the scripts, not just this one.

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+         required: false
+
+  def run
+    uri = URI("#{config[:url]}/manager/jmxproxy?get=#{config[:bean]}&att=#{config[:attribute]}")
```
FWIW, they _appear_ to be automatically being encoded. I've sent spaces and double-quotes and they are escaped appropriately. OTOH, I am only testing with Ruby 2.3 locally. If you know something about RestClient or its guts, especially in older versions of Ruby, that might need it, I'm happy to do so. It's just silly to do it twice is all. :)

majormoses replied:
ok, as long as we have looked at it and it's encoded that's fine by me. Agreed on double encoding. There are dozens of http methods and sometimes it's hard to remember which ones auto urlencode,

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+
+  def run
+    uri = URI("#{config[:url]}/manager/jmxproxy?get=#{config[:bean]}&att=#{config[:attribute]}")
+    fq_metric = config[:scheme] + '.' + (config[:metric] || config[:attribute])
```
What is Graphite/Sensu best practice?

majormoses replied:
From what I have seen in the metrics world most things just naturally are lowercase when pulling metric endpoints, so I think we should downcase. I only suggested it because I saw that the default would include mixed case.

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+    end
+
+    # Fetch attribute from Tomcat
+    Unirest.timeout(5) # 5s timeout
```
Nah, I was just following the lead of the other scripts in this repo

luyseyal commented on bin/metrics-tomcat-jmxproxy.rb:
```diff
@@ -0,0 +1,132 @@
+    Unirest.timeout(5) # 5s timeout
+    response = Unirest.get uri.to_s, auth: { user: @config[:account].to_s, password: @config[:password].to_s }
+
+    if response.code != 200
```
Indeed, Tomcat is often proxied. However, I think given the sensitivity of data from JMX and the fact that this relies on Basic Auth that automatically following redirects is a bad idea. Anyone using this script should be able to figure out what the correct URL they need to monitor is, right?! :) 

FWIW, Tomcat JMX Servlet appears to only return 200 even for errors such as invalid attribute.

luyseyal commented:
@majormoses I switched from URI to Addressable::URI which is supposed to be Better[tm] and which does require you to manually encode it (which I did -- `uri.normalize!`). I downcased the metric, made timeout a parameter, and updated the test and gemspec.

majormoses commented:
I have never used `Addressable::URI` but I checked and it does not have any extensions or dependencies that indicate the need of a compiler.

luyseyal commented:
Well, it was already a transitive dependency of the project before I added it to the gemspec. I could make it a compile dependency (that actually makes sense. Not sure why I stuck it in runtime...)

majormoses commented:
released: https://rubygems.org/gems/sensu-plugins-tomcat/versions/0.2.0

majormoses commented:
Can you expand on this:
> Well, it was already a transitive dependency of the project before I added it to the gemspec. I could make it a compile dependency (that actually makes sense. Not sure why I stuck it in runtime...)

I am not familiar with compile time dependencies in a gemspec. I see some issues asking for such a thing. Are you referring to using a ruby/c extension to install and install time?

luyseyal commented:
I could be saying the wrong words. I am used to Maven where you can have a compile dependency on a Java library that is also, by definition, a runtime dependency. In Maven, a runtime dependency is not required for it to compile so it is only run with tests and bundled with a jar/war/whatever. Since Ruby is interpreted rather than compiled (mostly), I suppose compile dependencies might not be the same thing.

majormoses commented:
In the ruby world since things normally do not require compilation there has never really been a distinction between compile time and runtime. That being said some do require a c compiler to build native extensions and people have asked for distinguishing them from other dependencies so that they may easily be removed post install as they are not actually required to run. The line gets fuzzier as time goes on as really I see software dependencies in far more ways than most package managers support:
- install time dependencies (I like to call these build dependencies)
- development dependencies (tools that really only someone developing or triaging problems would want)
- integration dependencies (very similar to dev deps but things that you might install in CI env that would not be installed locally for development)
- runtime dependencies
- system dependencies (things like a c compiler, or or headers provided by a system dependency, even if it is not able to be automatically installed at least called out in metadata)


Currently In the ruby world there are basically runtime and dev dependencies If it's needed to run then it is typically a runtime dependency if it is something for localdev, ci, etc, then it is a dev dependency. Eventually my goal is to have even system dependencies documented for each plugin that needs them in a uniform way. I am considering either gem metadata, readme, or by inspecting the bootstrap script for system integration testing which we are slowly rolling out across all the repos.
